### PR TITLE
fix(dev): keep too large video assets on device in gallery [AR-2918]

### PR DESCRIPTION
**Pull request by saleniuk**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/AR-2918" title="AR-2918" target="_blank">BugAR-2918</a>  Applause - Sending videos - App just discards / deletes video if 100MB exceeded
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
    - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
    - [x] contains a reference JIRA issue number like `SQPIT-764`
    - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
    - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

If the recorded video is bigger than the available limit, then this video is lost, because it's stored in the cache and the user cannot do anything with it (for instance trim it to be smaller and try to send again).

### Solutions

Cleaned up/refactored the code responsible for sending collected attachments by removing `AttachmentInnerState` which was a UI only layer where the attachment was stored as a state and then this state was instantly checked and the method in ViewModel to send a message was called. This was overcomplicated, the logic of checking the attachment size is in the ViewModel anyways. Right now, the selected attachment is passed directly to the ViewModel and everything is performed there, if needed we can simply add a state to be kept in that ViewModel.
Renamed some items and removed `ConversationViewState.kt` and moved all its elements to proper places, as it was a leftover after previous refactor which split the ViewModel into a few smaller ones.

Added a dialog to be shown when the size of attachment is too large. Added permissions to write to external storage for older Android versions. Implemented method to copy files to the external storage and add them to the gallery.

Needs releases with:

- https://github.com/wireapp/kalium/pull/1629

#### Test Coverage (Optional)

- [x] I have added automated test to this contribution

#### How to Test

Record a video longer than limit (25MB for guest, 100MB for team) and try to send that.

### Attachments (Optional)

https://user-images.githubusercontent.com/30429749/231392268-1e3d1d7a-4cfc-46cf-9e7e-af030ebea210.mp4

----
#### PR Post Submission Checklist for internal contributors (Optional)

- [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

- [x] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [1](https://10.10.124.134/job/wire-android-reloaded/job/PR-1653/1/) **failed**.

**Comment by AndroidBob** [comment]
Build [2](https://10.10.124.134/job/wire-android-reloaded/job/PR-1653/2/) **failed**.

**Comment by AndroidBob** [comment]
Build [3](https://10.10.124.134/job/wire-android-reloaded/job/PR-1653/3/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.1.1(6232406)(6232406)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.1.1(6232406)(6232406)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [4](https://10.10.124.134/job/wire-android-reloaded/job/PR-1653/4/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.1.1(6233984)(6233984)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.1.1(6233984)(6233984)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageComposerViewModel.kt [hunk 1]
```diff
@@ -396,5 +370,35 @@ class MessageComposerViewModel @Inject constructor(
         }
     }
 
-    fun provideTempCachePath(): Path = kaliumFileSystem.rootCachePath
+    @Suppress("MagicNumber")
+    fun attachmentPicked(attachmentUri: UriAsset) = viewModelScope.launch(dispatchers.io()) {
+        val tempCachePath = kaliumFileSystem.rootCachePath
+        val assetBundle = fileManager.getAssetBundleFromUri(attachmentUri.uri, tempCachePath)
+        if (assetBundle != null) {
+            // The max limit for sending assets changes between user types. Currently, is25MB for free users, and 100MB for
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageComposerViewModel.kt]: I wouldn't go too specific in this documentation on how big the limit is for each combination of asset vs user type. That's why we have the `GetAssetSizeLimitUseCase` for. Otherwise the moment we changed this limits in the use case, this documentation will become obsolete and we would probably forget to update it.
```suggestion
            // The max limit for sending assets changes between user and asset types. Check [GetAssetSizeLimitUseCase] class for more detailed information about the real limits. 
```

File: app/src/main/kotlin/com/wire/android/util/FileManager.kt [hunk 2]
```diff
@@ -82,6 +97,28 @@ class FileManager @Inject constructor(@ApplicationContext private val context: C
...
+        tempCachePath: Path,
+        dispatcher: DispatcherProvider = DefaultDispatcherProvider(),
+    ): AssetBundle? = withContext(dispatcher.io()) {
+            try {
+                val assetFileName = context.getFileName(attachmentUri) ?: throw IOException("The selected asset has an invalid name")
+                val fullTempAssetPath = "$tempCachePath/${UUID.randomUUID()}".toPath()
+                val mimeType = attachmentUri.getMimeType(context).orDefault(DEFAULT_FILE_MIME_TYPE)
+                val attachmentType = AttachmentType.fromMimeTypeString(mimeType)
+                val assetSize = if (attachmentType == AttachmentType.IMAGE) {
+                    attachmentUri.resampleImageAndCopyToTempPath(context, fullTempAssetPath)
```
Comment (gongracr) [inline app/src/main/kotlin/com/wire/android/util/FileManager.kt:111]: I know it is not on the scope of this PR, but maybe could you leave a comment highlighting the need of having to add a video resampling logic soon? that way we could drastically reduce as well the number of video assets hitting the max limit.

**Review by gongracr — APPROVED** [review]
Awesome work as usual 🙇 🍰  
I know that @trOnk12 was also touching bits of this message composer vm class, so maybe it would be great for him to take a peep in this PR so that he can adapt his changes more easily.

**Comment by AndroidBob** [comment]
Build [5](https://10.10.124.134/job/wire-android-reloaded/job/PR-1653/5/) **failed**.

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageComposerViewModel.kt [hunk 3]
```diff
@@ -396,5 +370,35 @@ class MessageComposerViewModel @Inject constructor(
...
+            // The max limit for sending assets changes between user and asset types.
+            // Check [GetAssetSizeLimitUseCase] class for more detailed information about the real limits.
+            val maxSizeLimitInBytes = getAssetSizeLimit(isImage = assetBundle.assetType == AttachmentType.IMAGE)
+            if (assetBundle.dataSize <= maxSizeLimitInBytes) {
+                sendAttachmentMessage(assetBundle)
+            } else {
+                if (attachmentUri.saveToDeviceIfInvalid) {
+                    with(assetBundle) { fileManager.saveToExternalMediaStorage(fileName, dataPath, dataSize, mimeType, dispatchers) }
+                }
+                val sizeOf1MB = 1024 * 1024
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageComposerViewModel.kt]: what about moving `sizeOf1MB` to companion object ?

File: app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageComposerViewModel.kt [hunk 4]
```diff
@@ -396,5 +370,35 @@ class MessageComposerViewModel @Inject constructor(
...
+                messageComposerViewState = messageComposerViewState.copy(
+                    assetTooLargeDialogState = AssetTooLargeDialogState.Visible(
+                        type = assetBundle.assetType,
+                        maxLimitInMB = maxSizeLimitInBytes.div(sizeOf1MB).toInt(),
+                        savedToDevice = attachmentUri.saveToDeviceIfInvalid
+                    )
+                )
+            }
+        } else {
+            onSnackbarMessage(ConversationSnackbarMessages.ErrorPickingAttachment)
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageComposerViewModel.kt:396]: Can you cover with test the case where `assetBundle` is null and we should show the snackbar ?

**Review by ohassine — APPROVED** [review]
👌

**Comment by AndroidBob** [comment]
Build [7](https://10.10.124.134/job/wire-android-reloaded/job/PR-1653/7/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-vv4.2.0(6284813)(6284813)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-vv4.2.0(6284813)(6284813)-staging-compat.apk)
